plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/DebugService.java

```diff
@@ -0,0 +1,216 @@
+
+import org.apache.log4j.Level;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.log4j.lf5.LogLevel;
+import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.MessageType;
+import org.eclipse.lsp4j.jsonrpc.services.JsonRequest;
+import org.eclipse.lsp4j.jsonrpc.services.JsonSegment;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.n4js.ide.xtext.server.DebugService.DebugServiceNullImpl;
+import org.eclipse.n4js.ide.xtext.server.build.ConcurrentIndex;
+import org.eclipse.n4js.utils.Strings;
+import org.eclipse.n4js.xtext.workspace.IWorkspaceConfigSnapshot;
+
+import com.google.inject.ImplementedBy;
+import com.google.inject.Inject;
+
+/**
+ * The {@link DebugService} uses a separate endpoint to avoid the infrastructure of other calls to ordinary LSP
+ * end-points and hence to increase robustness in case of errors in the source code.
+ */
+@JsonSegment("debug")
+@ImplementedBy(DebugServiceNullImpl.class)
+public interface DebugService {
+
+	/** Sets the client to communicate to */
+	void connect(LanguageClient client);
+
+	/** Sets the level of verbosity of Log4j. @see {@link LogLevel} */
+	@JsonRequest
+	CompletableFuture<Void> setVerboseLevel(String level);
+
+	/** Prints debug information on the output channel */
+	@JsonRequest
+	CompletableFuture<Void> printDebugInfo();
+
+	/** Default null implementation */
+	class DebugServiceNullImpl implements DebugService {
+		@Override
+		public void connect(LanguageClient client) {
+			// nothing to do
+		}
+
+		@Override
+		public CompletableFuture<Void> setVerboseLevel(String level) {
+			return CompletableFuture.failedFuture(new UnsupportedOperationException());
+		}
+
+		@Override
+		public CompletableFuture<Void> printDebugInfo() {
+			return CompletableFuture.failedFuture(new UnsupportedOperationException());
+		}
+	}
+
+	/** Default implementation */
+	class DebugServiceDefaultImpl implements DebugService {
+		private static final Logger LOG = LogManager.getLogger(DebugServiceDefaultImpl.class);
+
+		@Inject
+		private QueuedExecutorService executerService;
+
+		@Inject
+		private ConcurrentIndex fullIndex;
+
+		private LanguageClient client;
+
+		@Override
+		public void connect(LanguageClient _client) {
+			this.client = _client;
+		}
+
+		@Override
+		public CompletableFuture<Void> setVerboseLevel(String level) {
+			Level actualLevel = Level.toLevel(level.toUpperCase());
+			Logger.getRootLogger().setLevel(actualLevel);
+			LOG.info("log level set to " + Logger.getRootLogger().getLevel());
+			return CompletableFuture.completedFuture(null);
+		}
+
+		@Override
+		public CompletableFuture<Void> printDebugInfo() {
+			String msgText = getDebugInfo();
+			MessageParams message = new MessageParams();
+			message.setMessage(msgText);
+			message.setType(MessageType.Log);
+			client.logMessage(message);
+			return CompletableFuture.completedFuture(null);
+		}
+
+		private String getDebugInfo() {
```

Comment:
Maybe use protected visibility in this class to make it easier to reuse? Also `static` could be removed to allow overrides.